src/Interpreters/OptimizeDateOrDateTimeConverterWithPreimageVisitor.cpp

```diff
@@ -0,0 +1,188 @@
+ */
+ASTPtr generateOptimizedDateFilterAST(const String & comparator, const NameAndTypePair & column, const std::pair<Field, Field>& range)
+{
+    const DateLUTImpl & date_lut = DateLUT::instance();
+
+    const String & column_name = column.name;
+    String start_date_or_date_time;
+    String end_date_or_date_time;
+
+    if (isDateOrDate32(column.type.get()))
+    {
+        start_date_or_date_time = date_lut.dateToString(range.first.get<DateLUTImpl::Time>());
+        end_date_or_date_time = date_lut.dateToString(range.second.get<DateLUTImpl::Time>());
+    }
+    else if (isDateTime(column.type.get()) || isDateTime64(column.type.get()))
+    {
+        start_date_or_date_time = date_lut.timeToString(range.first.get<DateLUTImpl::Time>());
+        end_date_or_date_time = date_lut.timeToString(range.second.get<DateLUTImpl::Time>());
+    }
+    else [[unlikely]] return {};
+
+    if (comparator == "equals")
+    {
+        return makeASTFunction("and",
+                                makeASTFunction("greaterOrEquals",
+                                            std::make_shared<ASTIdentifier>(column_name),
+                                            std::make_shared<ASTLiteral>(start_date_or_date_time)
+                                            ),
+                                makeASTFunction("less",
+                                            std::make_shared<ASTIdentifier>(column_name),
+                                            std::make_shared<ASTLiteral>(end_date_or_date_time)
+                                            )
+                                );
+    }
+    else if (comparator == "notEquals")
+    {
+        return makeASTFunction("or",
+                                makeASTFunction("less",
+                                            std::make_shared<ASTIdentifier>(column_name),
+                                            std::make_shared<ASTLiteral>(start_date_or_date_time)
+                                            ),
+                                makeASTFunction("greaterOrEquals",
+                                            std::make_shared<ASTIdentifier>(column_name),
+                                            std::make_shared<ASTLiteral>(end_date_or_date_time)
+                                            )
+                                );
+    }
+    else if (comparator == "greater")
+    {
+        return makeASTFunction("greaterOrEquals",
+                    std::make_shared<ASTIdentifier>(column_name),
+                    std::make_shared<ASTLiteral>(end_date_or_date_time)
+                    );
+    }
+    else if (comparator == "lessOrEquals")
+    {
+        return makeASTFunction("less",
+                    std::make_shared<ASTIdentifier>(column_name),
+                    std::make_shared<ASTLiteral>(end_date_or_date_time)
+                    );
+    }
+    else // comparator == "less" || comparator == "greaterOrEquals"
```

Comment:
it's more safe to throw LOGICAL_ERROR exception when it's not less or greater or equal.
but it's up to you, just suggestion